src/HttpRequest.h

```diff
@@ -146,13 +149,38 @@ class HttpRequest: public Http::Message
 
     int imslen;
 
-    Ip::Address client_addr;
+    /// mark this request as initiated by Squid (rather than received on a client connection)
+    void selfInitiated();
+
+    /// whether this request was initiated by Squid (rather than received on a client connection)
+    bool isSelfInitiated() const { return selfInitiated_; }
+
+    /// supply Downloader-specific settings
+    void prepareForDownloader(Downloader *);
+
+    /// specify addresses manually when lacking client connection
+    void prepareForConnectionlessProtocol(const Ip::Address &fromAddr, const Ip::Address &localAddr);
+
+    /// the remote address of the client connection
+    const Ip::Address& clientAddr() const;
+
+    /// the local address of the client connection
+    const Ip::Address& myAddr() const;
+
+    /// the client connection manager of the underlying transaction, if any
+    CbcPointer<ConnStateData> &clientConnectionManager();
+
+    /// the client connection of the underlying transaction, if any
+    Comm::ConnectionPointer clientConnection() const;
 
 #if FOLLOW_X_FORWARDED_FOR
-    Ip::Address indirect_client_addr;
-#endif /* FOLLOW_X_FORWARDED_FOR */
+    /// the indirect client address
+    const Ip::Address& indirectClientAddr() const;
```

Comment:
When the indirect client address is unknown (including ignored), a reader would expect indirectClientAddr() to return an empty address or even throw an exception. However, indirectClientAddr() would often return direct client address (i.e. clientAddr()) in that case. The method name and useless description promise an indirect client address but the method sometimes returns a direct one. Sooner or later, a caller will be confused by this substitution. **If** the implementation is correct and all callers should be happy with the direct client address when the indirect one is unavailable, then I suggest renaming this method to furthestClientAddress().